feat(реализация кастомизации страницы блога)

baiboriev opened the pull request:
изменения совершаются через открывающую панель

baiboriev committed:
feat(реализация кастомизации страницы блога): изменения совершаются через открывающую панель

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,55 @@
+}: ArticleParamsFormProps) => {
+	const rootRef = useRef<HTMLDivElement>(null);
+
+	useOutsideClickClose({
```
Отлично:
Хорошо, что используете кастомный хук для оверлея

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,55 @@
+				className={clsx(styles.container, { [styles.container_open]: isOpen })}
+				ref={rootRef}>
+				<form className={styles.form} onSubmit={onSubmit}>
+					{children}
```
Надо исправить:
Компонент формы это форма, а не оболочка для отображения контента. Необходимо разметку формы реализовать внутри компонента.
Обратите внимание, что универсальный компонент для любой формы располагался бы в папке ui. При таком раскладе у вас все равно был бы дополнительный компонент, отвечающий за рендер контента именно этой формы.

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,55 @@
+			<aside
+				className={clsx(styles.container, { [styles.container_open]: isOpen })}
+				ref={rootRef}>
+				<form className={styles.form} onSubmit={onSubmit}>
```
**Отлично:**

Отлично, что обработчик сохранения данных вызывается при отправке формы.

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,33 +1,140 @@
 const domNode = document.getElementById('root') as HTMLDivElement;
 const root = createRoot(domNode);
 
 const App = () => {
```
**Можно лучше:**

Обычно корневой `index.tsx` держат максимально компактным. Стоит вынести компонент `App` в отдельный файл в папке `components/app/App.tsx`.

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,33 +1,140 @@
+	const [articleOptions, setArticleOptions] = useState({
+		...defaultArticleState,
+	});
+	const [isOpen, setIsOpen] = useState(false);
```
**Надо исправить:**

Состояние открытия формы необходимо создавать и хранить в самой форме. Каждый компонент сам должен отвечать за хранение и обработку своего состояния. Использовать подъем состояния формы в родительский компонент не имеет смысла для нашего приложения.

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,33 +1,140 @@
+	const [backgroundColor, setBackgroundColor] = useState(
+		defaultArticleState.backgroundColor
+	);
+	const [contentWidth, setContentWidth] = useState(
```
**Можно лучше:** 

Удобнее хранить все данные формы в одном стейте, а не создавать под каждое поле ввода свой стейт. Тогда единый обработчик для всех полей ввода будет выглядеть так:

```ts
const handleOnChange = (field: keyof ArticleStateType) => {
  return (value: OptionType) => {
    setState((prevState) => ({ ...prevState, [field]: value }));
  };
};
```

а его использование будет выглядеть так:

```ts
<Select
  title={'шрифт'}
  onChange={handleOnChange('fontFamilyOption')}
  selected={state.fontFamilyOption}
  options={fontFamilyOptions}
/>
```

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,33 +1,140 @@
+	const [fontFamilyOption, setFontFamilyOption] = useState(
+		defaultArticleState.fontFamilyOption
+	);
+	const [fontSizeOption, setFontSizeOption] = useState(
```
**Надо исправить:**

Каждый компонент сам должен отвечать за хранение и обработку своего состояния. Использовать подъем состояния в родительский компонент не имеет смысла для нашего приложения. Состояние формы нужно описать в самом компонент `ArticleParamsForm`, а не передавать его из `App`, т.к. это состояние описывает именно `ArticleParamsForm`. В состоянии `App` должны хранится только данные о текущем отображении статьи.

SergeyK2004 commented on src/index.tsx:
```diff
@@ -1,33 +1,140 @@
-		</main>
+		<>
+			<main
+				className={clsx(styles.main)}
```
Можно лучше:

Для одиночных селекторов не стоит использовать библиотеку `clsx`

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,55 @@
-				<form className={styles.form}>
+			<ArrowButton isOpen={isOpen} onClick={onClick} />
+			<aside
+				className={clsx(styles.container, { [styles.container_open]: isOpen })}
```
**Отлично:**

Здорово, что используете библиотеку `clsx`.

SergeyK2004 reviewed (changes requested):
Добрый день!  (нужно развернуть комментарий ↓)

У вас получилась отличная работа:

* Полностью реализован функционал.
* Сайдбар закрывается по клику вне сайдбара
* Обработчик события `submit` навешивается на форму, а не на клик по кнопке
* Для условного добавления стилей используется библиотека `clsx`
* Типизированы все пропсы


Но есть пара моментов, которые надо поправить:

- За состояние страницы отвечает компонент `App`, а за состояние формы компонент `ArticleParamsForm`. Нужно продумать логику передачи данных от `App` в форму и обратно.
- Компонент страницы не должен знать про состояние формы. Состояние формы надо перенести в компонент `ArticleParamsForm`. Туда же нужно перенести все обработчики.

Правки можно лучше остаются на ваше усмотрение:

- Обычно корневой `index.tsx` держат максимально компактным. Стоит вынести компонент `App` в отдельный файл в папке `components/app/App.tsx`.
- Для одиночных селекторов не стоит использовать библиотеку `clsx` (см. компоненты `Article` и `App`).
- Использовать один обработчик для событий `onChange` на элементах формы

Успешного рефакторинга!

baiboriev committed:
fix: компонент App был вынесен в отдельный файл, изменения логики передачи данных

SergeyK2004 commented on src/components/app/App.tsx:
```diff
@@ -0,0 +1,34 @@
+export const App = () => {
+	const [articleState, setArticleState] = useState({ ...defaultArticleState });
+
+	const onChange = (newData: ArticleStateType) => {
```
Не нужно создавать функцию вызывающую функцию. Передавайте в форму сразу функцию изменяющую состояние `setArticleState`

SergeyK2004 commented on src/components/app/App.tsx:
```diff
@@ -0,0 +1,34 @@
+					'--bg-color': articleState.backgroundColor.value,
+				} as CSSProperties
+			}>
+			<ArticleParamsForm articleState={articleState} onChange={onChange} />
```
Не стоит передавать в дочерний компонент лишние пропсы. Достаточно только передать функцию изменения состояния

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,122 @@
+	const [isOpen, setIsOpen] = useState(false);
+	const asideRef = useRef<HTMLDivElement>(null);
+
+	const [customizerState, setCustomizerState] = useState({
```
**Надо исправить:**

Настройки статьи по умолчанию описаны в объекте `defaultArticleState` в файле `constants/articleProps.ts` стартеркита. В проекте нужно использовать значения из этого объекта как настройки по умолчанию, как для самой статьи, так и как начальные значения полей форм.

SergeyK2004 commented on src/components/article-params-form/ArticleParamsForm.tsx:
```diff
@@ -1,16 +1,122 @@
+		onChange: setIsOpen,
+	});
+
+	const handleOnChange = (field: keyof ArticleStateType) => {
```
Отлично:
Здорово, что один обработчик на все поля формы.

SergeyK2004 reviewed (changes requested):
Кайрат, добрый день.
Очень хорошие правки. Приложение отличное. 
Осталось совсем немного. Посмотри работу со стейтом статьи.
Исправь замечания в коде и работа будет принята.

baiboriev committed:
fix(app.tsx articleparamsform.tsx): исправлены компоненты App и ArticleParamsForm

В ArticleParamsForm значения по умолчанию берутся в объекте defaultArticleState, исправлен
ArticleParamsFormProps. В App удалена лишняя функция, исправлена передача данных в дочерний
компонент.

fix #1

SergeyK2004 reviewed (approved):
Отличная работа.
Замечаний больше нет.
Поздравляю со сдачей проекта.
Желаю успехов в дальнейшем обучении.
